### ISSUE TITLE
Connection blocked by 3rd party DNS isn't shown in connection history

### ISSUE THREAD

**Issue by northys** [safing/portmaster#540]
If this is expected behaviour just close it but it seems weird to me that connections blocked by portmaster are shown whereas those blocked by 3rd party aren't there considering portmaster knows it happened because it  handled the query and got zeroip responce which led to blocking the connection. This looks like inconsistency to me.

---

**What happened**:

I played around with nextdns to see whether it blocks more than portmaster itself and I noticed that the connection is not shown in the connection history when 3rd party DNS blocks it. First I was confused, because analytics.google.com didn't work and I had it enabled for firefox because sometimes I need to open the dashboard so I block only the analytics.js using ublock. Also I couldn't find the connection in portmaster at all so I though something bad is happening after latest upgrade but then I relaized I switched to nextdns as main DNS yesterday...

**What did you expect to happen?**:

I would expect to see that connection (even there was no connection but my laptop as a client side wanted to open it) when DNS decides to reply with zeroip.

**How did you reproduce it?**:

Use nextdns and open blocked domain.

**Comment by dhaavi** [comment]
I just saw something related in a log from another of your issues.
The connections are wrongly marked as "failed", not as blocked when blocked by a 3rd party. Already have a fix.

Can you see a "yellow dot" failed connection for the DNS request in logs?
Oh yes, and that's the thing with the 3rd party dns, these now might land in the System DNS Client app due to how that works - explained in other issue - #541.

Oh, and we have blog post about that: https://safing.io/blog/2021/03/23/attributing-dns-requests-on-windows/
It was written for Windows, but systemd-resolved works very similar.

**Comment by northys** [comment]
> Can you see a "yellow dot" failed connection for the DNS request in network monitor?

When I `dig` it there is green connection which expands to another line with yellow blocked DNS request. Unfortunately when I open the same domain I've digged using firefox the connection is also blocked but nothing is shown in portmaster at all.

I would share the firefox debug data, but I have too many websites opened I don't wanna expose. Also I couldn't see the domain `iam.northys.cz` mentioned there. Also the System DNS Client did not pop in so I expect that firefox queries through portmaster directly.

first two (from bottom :)) are from dig, the last one (from bottom :)) is from firefox.
![Screenshot from 2022-02-25 19-33-04](https://user-images.githubusercontent.com/2878126/155769304-a2ada2d0-2cee-4e69-b9d4-7c15061d82bb.png)

**Comment by dhaavi** [comment]
There were some improvements in v0.8.5. Can you check again?

Blocked DNS queries should always show up. Successful ones are normally "merged into" the connection, and are only shown when the request was not served from cache - else DNS requests would needlessly spam the interface.

**Comment by northys** [comment]
@dhaavi we talked about that dbus thing. Would the blocked connection by remote DNS also show up in history if firefox did not open connection to the resolved IP? I'm talking about the portmaster matching magic when resolves are not handled by DNS protocol. That would explain that this works for dig and not for firefox as I stated in first post.

---

I'll try it again (maybe today) later but I think I have 0.8.5. for quite a while.

**Comment by dhaavi** [comment]
Blocked DNS requests are always visible. Allowed DNS requests are "merged" into the follow-up connection within one second (iirc) and are only logged themselves when it was not served from cache, to reduce superfluous connections. If queries go over dbus and there is _no_ follow up connection, these DNS requests would show up in the System DNS Client instead.

**Comment by northys** [comment]
works

![Screenshot from 2022-04-11 21-52-19](https://user-images.githubusercontent.com/2878126/162818825-46888ba5-e449-4c02-8e72-6e601b2b96a3.png)
